File: app/src/main/java/com/jawnnypoo/geotune/BaseActivity.java
package com.jawnnypoo.geotune;

import android.content.Intent;
import android.media.RingtoneManager;
import android.net.Uri;
import android.support.v7.app.AppCompatActivity;

import com.jawnnypoo.geotune.data.GeoTune;

import java.util.ArrayList;

import timber.log.Timber;

/**
 * Base activity for all others to derive
 */
public class BaseActivity extends AppCompatActivity {

    protected static final int REQUEST_GEOFENCE = 819;
    protected static final int REQUEST_AUDIO = 67;
    protected static final int REQUEST_NOTIFICATION = 68;

    public static final String EXTRA_GEOTUNE = "extra_geotune";

    protected void navigateToMap(int[] location, ArrayList<GeoTune> geoTunes) {
        startActivityForResult(GeoMapActivity.newIntent(this, location, geoTunes), REQUEST_GEOFENCE);
        overridePendingTransition(R.anim.do_nothing, R.anim.do_nothing);
    }

    protected void navigateToMap(int[] location, ArrayList<GeoTune> geoTunes, GeoTune startingGeoTune) {
        startActivityForResult(GeoMapActivity.newIntent(this, location, geoTunes, startingGeoTune),
                REQUEST_GEOFENCE);
        overridePendingTransition(R.anim.do_nothing, R.anim.do_nothing);
    }

    protected void navigateToAbout() {
        startActivity(AboutActivity.newInstance(this));
        overridePendingTransition(R.anim.fade_in, R.anim.do_nothing);
    }

    protected void chooseNotification() {
        Intent intent = new Intent(RingtoneManager.ACTION_RINGTONE_PICKER);
        intent.putExtra(RingtoneManager.EXTRA_RINGTONE_TYPE, RingtoneManager.TYPE_NOTIFICATION);
        intent.putExtra(RingtoneManager.EXTRA_RINGTONE_TITLE, "Select Tone");
        intent.putExtra(RingtoneManager.EXTRA_RINGTONE_EXISTING_URI, (Uri) null);
        this.startActivityForResult(intent, REQUEST_NOTIFICATION);
    }

    protected void chooseRingtone() {
        Intent intent = new Intent(RingtoneManager.ACTION_RINGTONE_PICKER);
        intent.putExtra(RingtoneManager.EXTRA_RINGTONE_TYPE, RingtoneManager.TYPE_RINGTONE);
        intent.putExtra(RingtoneManager.EXTRA_RINGTONE_TITLE, "Select Tone");
        intent.putExtra(RingtoneManager.EXTRA_RINGTONE_EXISTING_URI, (Uri) null);
        this.startActivityForResult(intent, REQUEST_NOTIFICATION);
    }

    protected void chooseMedia() {
        Timber.d("ChooseAudio called");
        Intent intent = new Intent();
        intent.setType("audio/*");
        intent.setAction(Intent.ACTION_GET_CONTENT);
        startActivityForResult(Intent.createChooser(intent, "Select Audio "), REQUEST_AUDIO);
    }

    @Override
    public void finish() {
        super.finish();
        if (this instanceof MainActivity) {

        } else {
            overridePendingTransition(R.anim.do_nothing, R.anim.fade_out);
        }
    }
}

File: app/src/main/java/com/jawnnypoo/geotune/activity/BaseActivity.kt
package com.jawnnypoo.geotune.activity

import android.content.Intent
import android.media.RingtoneManager
import android.net.Uri
import android.support.v7.app.AppCompatActivity
import com.jawnnypoo.geotune.R
import com.jawnnypoo.geotune.data.GeoTune
import timber.log.Timber
import java.util.*

/**
 * Base activity for all others to derive
 */
open class BaseActivity : AppCompatActivity() {

    companion object {

        const val REQUEST_GEOFENCE = 819
        const val REQUEST_AUDIO = 67
        const val REQUEST_NOTIFICATION = 68

        val EXTRA_GEOTUNE = "extra_geotune"
    }

    protected fun navigateToMap(location: IntArray, geoTunes: List<GeoTune>) {
        startActivityForResult(GeoMapActivity.newIntent(this, location, geoTunes), REQUEST_GEOFENCE)
        overridePendingTransition(R.anim.do_nothing, R.anim.do_nothing)
    }

    protected fun navigateToMap(location: IntArray, geoTunes: List<GeoTune>, startingGeoTune: GeoTune) {
        startActivityForResult(GeoMapActivity.newIntent(this, location, geoTunes, startingGeoTune),
                REQUEST_GEOFENCE)
        overridePendingTransition(R.anim.do_nothing, R.anim.do_nothing)
    }

    protected fun navigateToAbout() {
        startActivity(AboutActivity.newInstance(this))
        overridePendingTransition(R.anim.fade_in, R.anim.do_nothing)
    }

    protected fun chooseNotification() {
        val intent = Intent(RingtoneManager.ACTION_RINGTONE_PICKER)
        intent.putExtra(RingtoneManager.EXTRA_RINGTONE_TYPE, RingtoneManager.TYPE_NOTIFICATION)
        intent.putExtra(RingtoneManager.EXTRA_RINGTONE_TITLE, "Select Tone")
        intent.putExtra(RingtoneManager.EXTRA_RINGTONE_EXISTING_URI, null as Uri?)
        this.startActivityForResult(intent, REQUEST_NOTIFICATION)
    }

    protected fun chooseRingtone() {
        val intent = Intent(RingtoneManager.ACTION_RINGTONE_PICKER)
        intent.putExtra(RingtoneManager.EXTRA_RINGTONE_TYPE, RingtoneManager.TYPE_RINGTONE)
        intent.putExtra(RingtoneManager.EXTRA_RINGTONE_TITLE, "Select Tone")
        intent.putExtra(RingtoneManager.EXTRA_RINGTONE_EXISTING_URI, null as Uri?)
        this.startActivityForResult(intent, REQUEST_NOTIFICATION)
    }

    protected fun chooseMedia() {
        Timber.d("ChooseAudio called")
        val intent = Intent()
        intent.type = "audio/*"
        intent.action = Intent.ACTION_GET_CONTENT
        startActivityForResult(Intent.createChooser(intent, "Select Audio "), REQUEST_AUDIO)
    }

    override fun finish() {
        super.finish()
        if (this is MainActivity) {

        } else {
            overridePendingTransition(R.anim.do_nothing, R.anim.fade_out)
        }
    }
}
